// libmonad/index.js
// Generated by CoffeeScript 1.6.1
(function() {
  var cont_m, cont_t, domonad, drop_while, first, identity, identity_m, is_function, is_null, lift_async, lift_sync, logger_m, logger_t, maybe_m, maybe_t, _ref,
    __slice = [].slice;

  _ref = require('libprotein'), identity = _ref.identity, first = _ref.first, drop_while = _ref.drop_while, is_function = _ref.is_function;

  is_null = function() {
    var v;
    v = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    if (v.length === 0) {
      return null;
    } else {
      return v[0] === null;
    }
  };

  domonad = function(_arg, functions, init_value) {
    var bind, f0, result;
    result = _arg.result, bind = _arg.bind;
    f0 = bind(result(init_value), functions[0]);
    return ([f0].concat(functions.slice(1))).reduce(function(a, b) {
      return bind(a, b);
    });
  };

  identity_m = function() {
    return {
      result: identity,
      bind: function(mv, f) {
        return f(mv);
      }
    };
  };

  maybe_m = function(_arg) {
    var is_error;
    is_error = _arg.is_error;
    return {
      zero: function() {
        return is_error();
      },
      result: function(v) {
        return v;
      },
      bind: function(mv, f) {
        if (is_error(mv)) {
          return mv;
        } else {
          return f(mv);
        }
      },
      plus: function() {
        var mvs;
        mvs = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return first(drop_while(is_error(mvs)));
      }
    };
  };

  maybe_t = function(inner) {
    return {
      result: function(v) {
        return inner.result(v);
      },
      bind: function(mv, f) {
        if (is_error(mv)) {
          return mv;
        } else {
          return inner.bind(mv, f);
        }
      }
    };
  };

  logger_m = function(log_fn) {
    return {
      result: function(v) {
        log_fn("Got value:", v);
        return v;
      },
      bind: function(mv, f) {
        var r;
        log_fn("Going to call f(mv):", f, mv);
        r = f(mv);
        return log_fn("Got result:", r);
      }
    };
  };

  logger_t = function(inner, log_fn) {
    return {
      result: function(v) {
        var r;
        log_fn("Got value:", v);
        r = inner.result(v);
        log_fn("Got inner monad's result value:", r);
        return r;
      },
      bind: function(mv, f) {
        var r;
        log_fn("Going to call f(mv):", f, mv);
        r = inner.bind(mv, f);
        log_fn("Got result:", r);
        return r;
      }
    };
  };

  cont_m = function() {
    return {
      result: function(v) {
        return function(c) {
          return c(v);
        };
      },
      bind: function(mv, f) {
        return function(c) {
          return mv((function(v) {
            return (f(v))(c);
          }));
        };
      }
    };
  };

  cont_t = function(inner) {
    return {
      result: function(v) {
        return function(c) {
          return c(inner.result(v));
        };
      },
      bind: function(mv, f) {
        return function(c) {
          var get_h;
          get_h = function(v) {
            var inner_bind_res;
            inner_bind_res = inner.bind(v, f);
            if (is_function(inner_bind_res)) {
              return inner_bind_res;
            } else {
              return function(c) {
                return c(inner_bind_res);
              };
            }
          };
          return mv((function(v) {
            return (get_h(v))(c);
          }));
        };
      }
    };
  };

  lift_sync = function(arity, f) {
    ' Lifts a function:\nf: arg1 -> ... -> argN\nto a function:\nf1: (arg1 -> ... -> argN) -> cont';    return function() {
      var args;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      return function(c) {
        var res;
        res = f.apply(null, args.slice(0, arity));
        return c(res);
      };
    };
  };

  lift_async = function(arity, f) {
    ' Lifts a function:\nf: arg1 -> ... -> argN -> cb\nto a function:\nf1: (arg1 -> ... argN) -> cont';    return function() {
      var args;
      args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
      return function(c) {
        return f.apply(null, args.slice(0, arity - 1).concat([c]));
      };
    };
  };

  module.exports = {
    domonad: domonad,
    identity_m: identity_m,
    maybe_m: maybe_m,
    maybe_t: maybe_t,
    cont_m: cont_m,
    cont_t: cont_t,
    logger_m: logger_m,
    logger_t: logger_t,
    lift_sync: lift_sync,
    lift_async: lift_async,
    is_null: is_null
  };

}).call(this);
