// libprotein/index.js
// Generated by CoffeeScript 1.6.1
(function() {
  var __slice = [].slice;

  module.exports = {
    partial: function() {
      var f, partial_args;
      f = arguments[0], partial_args = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
      return function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return f.apply(null, partial_args.concat(args));
      };
    },
    complement: function(f) {
      return function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return !(f.apply(null, args));
      };
    },
    compose2: function(f, g) {
      return function() {
        var args;
        args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
        return f(g.apply(null, args));
      };
    },
    first: function(s) {
      return s[0];
    },
    identity: function(x) {
      return x;
    },
    drop_while: function(f, s) {
      var i, _i, _len;
      for (_i = 0, _len = s.length; _i < _len; _i++) {
        i = s[_i];
        if (!(f(i))) {
          return i;
        }
      }
    },
    is_function: function(v) {
      return typeof v === 'function';
    },
    is_array: function(v) {
      return Array.isArray(v);
    },
    is_object: function(v) {
      if (is_array(v)) {
        return false;
      } else {
        return v instanceof {}.constructor;
      }
    },
    bool: function(v) {
      if (is_array(v)) {
        return !!v.length;
      } else if (is_object(v)) {
        return !!(Object.keys(v).length);
      } else {
        return !!v;
      }
    }
  };

}).call(this);
